File: apps/api/scripts/fix-morning-shift-import.ts
/**
 * Fix Morning Shift Sem 3 import Excel and save with ExcelJS (ERP-compatible).
 *
 *   npx ts-node --transpile-only scripts/fix-morning-shift-import.ts
 *   npx ts-node --transpile-only scripts/fix-morning-shift-import.ts --validate
 */
import * as fs from 'fs';
import * as path from 'path';
import ExcelJS from 'exceljs';
import { NestFactory } from '@nestjs/core';
import { parseFlexibleDate } from '../src/common/utils/parse-flexible-date';
import { DBC_SEM3_COURSE_TITLES } from '../src/modules/academic-engine/domain/dbc-morning-sem3-catalog';
import { AppModule } from '../src/app.module';
import { PrismaService } from '../src/database/prisma.service';
import { StudentImportService } from '../src/modules/students/import/student-import.service';

/**
 * Titles as stored on Course rows for Morning Shift Sem 3 import resolution.
 * Desktop Publishing / Computerized Accounting use the live course-master titles
 * (VTC: 243.2 / VTC: 243.3) so import dropdown resolution matches exactly.
 */
const IMPORT_PAPER_TITLES: Record<string, string> = {
  ...DBC_SEM3_COURSE_TITLES,
  'VTC-243.2': 'Desktop Publishing -I',
  'VTC-243.3': 'Computerized Accounting-I',
};

/** Morning Sem 3 VTC pool — includes papers students actually take. */
const MORNING_IMPORT_VTC_CODES = [
  'VTC-240.3',
  'VTC-241.2',
  'VTC-243.2',
  'VTC-243.3',
  'VTC-244.2',
  'VTC-245.3',
  'VTC-246.1',
  'VTC-248.1',
] as const;

const MORNING_VTC_TITLES = new Set(
  MORNING_IMPORT_VTC_CODES.map((code) => IMPORT_PAPER_TITLES[code]),
);

const CANONICAL_PAPER_TITLES = [
  ...Object.values(IMPORT_PAPER_TITLES),
  ...Object.values(DBC_SEM3_COURSE_TITLES),
];

const SOURCE = path.join(
  process.env.USERPROFILE ?? '',
  'OneDrive',
  'Desktop',
  'Import Live 1-3-5',
  'Morning Shift',
  'MORNING SHIFT IMPORT DATA 2026.xlsx',
);
const OUTPUT = SOURCE.replace('.xlsx', ' - READY.xlsx');
const REPORT = SOURCE.replace('.xlsx', ' - VALIDATION REPORT.txt');
const validate = process.argv.includes('--validate');

function normalizePaperLabel(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

function canonicalizePaperTitle(value: ExcelJS.CellValue): string | null {
  const raw = cellText(value);
  if (!raw) return null;
  const normalized = normalizePaperLabel(raw);
  const exact = CANONICAL_PAPER_TITLES.find(
    (title) => normalizePaperLabel(title) === normalized,
  );
  if (exact) return exact;

  const partial = CANONICAL_PAPER_TITLES.filter((title) => {
    const candidate = normalizePaperLabel(title);
    return candidate.includes(normalized) || normalized.includes(candidate);
  });
  if (partial.length === 1) return partial[0];

  const aliasFixes: Record<string, string> = {
    'baking and confectionary i': IMPORT_PAPER_TITLES['VTC-246.1'],
    'baking and confectionery i': IMPORT_PAPER_TITLES['VTC-246.1'],
    'bee keeping i': IMPORT_PAPER_TITLES['VTC-240.3'],
    'bee keeping - i': IMPORT_PAPER_TITLES['VTC-240.3'],
    'mushroom cultivation i': IMPORT_PAPER_TITLES['VTC-241.2'],
    'guitar i': IMPORT_PAPER_TITLES['VTC-245.3'],
    'event management i': IMPORT_PAPER_TITLES['VTC-244.2'],
    'desktop publishing': IMPORT_PAPER_TITLES['VTC-243.2'],
    'desktop publishing i': IMPORT_PAPER_TITLES['VTC-243.2'],
    'computerized accounting': IMPORT_PAPER_TITLES['VTC-243.3'],
    'computerized accounting i': IMPORT_PAPER_TITLES['VTC-243.3'],
  };
  if (aliasFixes[normalized]) return aliasFixes[normalized];

  return raw;
}

const PROGRAM_SEM3_DEFAULTS: Record<
  string,
  { mdc: string; sec: string; vtc: string }
> = {
  'BA-EDU': {
    mdc: IMPORT_PAPER_TITLES['MDC-211'],
    sec: IMPORT_PAPER_TITLES['SEC-230'],
    vtc: IMPORT_PAPER_TITLES['VTC-241.2'],
  },
  'BA-ENG': {
    mdc: IMPORT_PAPER_TITLES['MDC-211'],
    sec: IMPORT_PAPER_TITLES['SEC-230'],
    vtc: IMPORT_PAPER_TITLES['VTC-246.1'],
  },
  'BA-POL': {
    mdc: IMPORT_PAPER_TITLES['MDC-212'],
    sec: IMPORT_PAPER_TITLES['SEC-230'],
    vtc: IMPORT_PAPER_TITLES['VTC-245.3'],
  },
  'BA-ECO': {
    mdc: IMPORT_PAPER_TITLES['MDC-212'],
    sec: IMPORT_PAPER_TITLES['SEC-230'],
    vtc: IMPORT_PAPER_TITLES['VTC-244.2'],
  },
  'BA-GAR': {
    mdc: IMPORT_PAPER_TITLES['MDC-211'],
    sec: IMPORT_PAPER_TITLES['SEC-230'],
    vtc: IMPORT_PAPER_TITLES['VTC-245.3'],
  },
  'BA-HIS': {
    mdc: IMPORT_PAPER_TITLES['MDC-211'],
    sec: IMPORT_PAPER_TITLES['SEC-230'],
    vtc: IMPORT_PAPER_TITLES['VTC-241.2'],
  },
  'BA-SOC': {
    mdc: IMPORT_PAPER_TITLES['MDC-211'],
    sec: IMPORT_PAPER_TITLES['SEC-230'],
    vtc: IMPORT_PAPER_TITLES['VTC-241.2'],
  },
};

const KNOWN_DOB_FIXES: Record<string, string> = {
  '22.06.32007': '22.06.2007',
  '14.1.02005': '14.10.2005',
  '30.102005': '30.10.2005',
  '1.0702004': '01.07.2004',
};

function cellText(value: ExcelJS.CellValue): string {
  if (value == null) return '';
  if (value instanceof Date) return value.toISOString();
  if (typeof value === 'object') {
    if ('result' in value && value.result != null)
      return cellText(value.result);
    if ('text' in value && value.text) return String(value.text);
    if ('richText' in value && Array.isArray(value.richText)) {
      return value.richText.map((part) => part.text ?? '').join('');
    }
  }
  return String(value).trim();
}

function isBlank(value: ExcelJS.CellValue): boolean {
  const text = cellText(value);
  return !text || text.toLowerCase() === 'nan';
}

function repairDobText(raw: string): string {
  if (KNOWN_DOB_FIXES[raw]) return KNOWN_DOB_FIXES[raw];
  const longYear = raw.match(/^(\d{1,2})\.(\d{1,2})\.(\d{4,})$/);
  if (longYear) {
    const [, day, month, yearPart] = longYear;
    return `${day.padStart(2, '0')}.${month.padStart(2, '0')}.${yearPart.slice(-4)}`;
  }
  const missingDot = raw.match(/^(\d{1,2})\.(\d{2})(\d{4})$/);
  if (missingDot) {
    const [, day, month, year] = missingDot;
    if (Number(month) >= 1 && Number(month) <= 12) {
      return `${day.padStart(2, '0')}.${month.padStart(2, '0')}.${year}`;
    }
  }
  return raw;
}

function normalizeDate(value: ExcelJS.CellValue): string | null {
  if (value instanceof Date) {
    return parseFlexibleDate(value);
  }
  const raw = cellText(value);
  if (!raw) return null;
  return parseFlexibleDate(repairDobText(raw));
}

function normalizeGender(value: ExcelJS.CellValue): string | null {
  const raw = cellText(value);
  if (!raw) return null;
  const upper = raw.toUpperCase();
  if (upper === 'MALE' || upper === 'M') return 'Male';
  if (upper === 'FEMALE' || upper === 'F') return 'Female';
  if (upper === 'OTHER' || upper === 'O') return 'Other';
  return raw;
}

function normalizeCategory(value: ExcelJS.CellValue): string | null {
  const raw = cellText(value);
  if (!raw) return null;
  const upper = raw.toUpperCase();
  if (upper === 'GEN' || upper === 'GENERAL') return 'GENERAL';
  return upper;
}

function rollToEmail(roll: string): string {
  const slug = roll.toLowerCase().replace(/[^a-z0-9]/g, '');
  return `${slug}@student.donboscocollege.ac.in`;
}

function headerMap(sheet: ExcelJS.Worksheet): Map<string, number> {
  const map = new Map<string, number>();
  sheet.getRow(1).eachCell((cell, col) => {
    const header = cellText(cell.value);
    if (header) map.set(header, col);
  });
  return map;
}

function setCell(row: ExcelJS.Row, col: number | undefined, value: string) {
  if (!col) return;
  row.getCell(col).value = value;
}

async function main() {
  if (!fs.existsSync(SOURCE)) {
    throw new Error(`Source file not found: ${SOURCE}`);
  }

  const workbook = new ExcelJS.Workbook();
  await workbook.xlsx.readFile(SOURCE);
  const sheet = workbook.getWorksheet('Students');
  if (!sheet) throw new Error('Students sheet not found');

  const columns = headerMap(sheet);
  const required = [
    'Registration Number',
    'Roll Number',
    'Full Name',
    'Email Address',
    'Date of Birth',
    'Programme',
    'MDC (Sem 3)',
    'AEC (Sem 3)',
    'SEC (Sem 3)',
    'VTC',
    'Student Status',
    'Gender',
    'Category',
    'Admission Date',
  ];
  for (const header of required) {
    if (!columns.has(header)) {
      throw new Error(`Missing expected column: ${header}`);
    }
  }

  const stats = {
    rows: 0,
    registrationFilled: 0,
    dobFixed: 0,
    emailGenerated: 0,
    renewalAssigned: 0,
    vtcRemapped: 0,
    statusFilled: 0,
    duplicatesRemoved: 0,
    issues: [] as string[],
  };

  const dataStartRow = cellText(
    sheet.getRow(2).getCell(columns.get('Registration Number')!).value,
  ).startsWith('College')
    ? 3
    : 2;

  for (let rowNumber = dataStartRow; rowNumber <= sheet.rowCount; rowNumber++) {
    const row = sheet.getRow(rowNumber);
    const roll = cellText(row.getCell(columns.get('Roll Number')!).value);
    const name = cellText(row.getCell(columns.get('Full Name')!).value);
    if (!roll && !name) continue;
    stats.rows += 1;

    const regCol = columns.get('Registration Number')!;
    if (isBlank(row.getCell(regCol).value) && roll) {
      row.getCell(regCol).value = roll;
      stats.registrationFilled += 1;
    }

    const dobCol = columns.get('Date of Birth')!;
    const dobBefore = cellText(row.getCell(dobCol).value);
    const dobNormalized = normalizeDate(row.getCell(dobCol).value);
    if (dobNormalized) {
      if (dobBefore !== dobNormalized) {
        row.getCell(dobCol).value = dobNormalized;
        stats.dobFixed += 1;
      }
    } else if (isBlank(row.getCell(dobCol).value)) {
      stats.issues.push(`Row ${rowNumber}: missing Date of Birth (${name})`);
    } else {
      stats.issues.push(
        `Row ${rowNumber}: unparseable DOB '${dobBefore}' (${name})`,
      );
    }

    const admCol = columns.get('Admission Date')!;
    const admNormalized = normalizeDate(row.getCell(admCol).value);
    if (admNormalized) row.getCell(admCol).value = admNormalized;

    const emailCol = columns.get('Email Address')!;
    if (isBlank(row.getCell(emailCol).value)) {
      if (roll) {
        row.getCell(emailCol).value = rollToEmail(roll);
        stats.emailGenerated += 1;
      } else {
        stats.issues.push(`Row ${rowNumber}: missing email and roll (${name})`);
      }
    } else {
      row.getCell(emailCol).value = cellText(
        row.getCell(emailCol).value,
      ).toLowerCase();
    }

    const programme = cellText(row.getCell(columns.get('Programme')!).value);
    const defaults = PROGRAM_SEM3_DEFAULTS[programme];
    if (isBlank(row.getCell(columns.get('MDC (Sem 3)')!).value) && defaults) {
      setCell(row, columns.get('MDC (Sem 3)'), defaults.mdc);
      setCell(row, columns.get('SEC (Sem 3)'), defaults.sec);
      setCell(row, columns.get('VTC'), defaults.vtc);
      stats.renewalAssigned += 1;
    }

    if (isBlank(row.getCell(columns.get('AEC (Sem 3)')!).value)) {
      setCell(row, columns.get('AEC (Sem 3)'), IMPORT_PAPER_TITLES['AEC-222']);
    }

    for (const header of [
      'MDC (Sem 3)',
      'AEC (Sem 3)',
      'SEC (Sem 3)',
      'VTC',
    ] as const) {
      const col = columns.get(header);
      if (!col) continue;
      let canonical = canonicalizePaperTitle(row.getCell(col).value);
      if (header === 'VTC' && canonical) {
        const inMorningPool = [...MORNING_VTC_TITLES].some(
          (title) =>
            normalizePaperLabel(title) === normalizePaperLabel(canonical!),
        );
        if (!inMorningPool) {
          const replacement = defaults?.vtc ?? IMPORT_PAPER_TITLES['VTC-248.1'];
          if (
            normalizePaperLabel(canonical) !== normalizePaperLabel(replacement)
          ) {
            stats.vtcRemapped += 1;
            stats.issues.push(
              `Row ${rowNumber}: VTC "${canonical}" not in Morning pool → "${replacement}" (${name})`,
            );
          }
          canonical = replacement;
        }
      }
      if (canonical) row.getCell(col).value = canonical;
    }

    const fatherCol = columns.get("Father's Name");
    const motherCol = columns.get("Mother's Name");
    const guardianCol = columns.get('Guardian Name');
    const guardian = guardianCol
      ? cellText(row.getCell(guardianCol).value)
      : '';
    if (fatherCol && isBlank(row.getCell(fatherCol).value)) {
      row.getCell(fatherCol).value = guardian || 'Not Provided';
    }
    if (motherCol && isBlank(row.getCell(motherCol).value)) {
      row.getCell(motherCol).value = guardian || 'Not Provided';
    }

    const tribeCol = columns.get('Tribe / Race');
    if (tribeCol) {
      const tribe = cellText(row.getCell(tribeCol).value).toUpperCase();
      if (tribe === 'RABHA' || tribe === 'SHIL') {
        row.getCell(tribeCol).value = '';
      }
    }

    const statusCol = columns.get('Student Status')!;
    if (isBlank(row.getCell(statusCol).value)) {
      row.getCell(statusCol).value = 'STUDYING';
      stats.statusFilled += 1;
    }

    const genderCol = columns.get('Gender')!;
    const gender = normalizeGender(row.getCell(genderCol).value);
    if (gender) row.getCell(genderCol).value = gender;

    const categoryCol = columns.get('Category')!;
    const category = normalizeCategory(row.getCell(categoryCol).value);
    if (category) row.getCell(categoryCol).value = category;

    const mobileCol = columns.get('Student Mobile Number');
    if (mobileCol && isBlank(row.getCell(mobileCol).value)) {
      stats.issues.push(`Row ${rowNumber}: missing mobile number (${name})`);
    }
  }

  const uniRollCol = columns.get('University Roll Number');
  if (uniRollCol) {
    const winners = new Map<
      string,
      { rowNumber: number; score: number; name: string }
    >();
    for (
      let rowNumber = dataStartRow;
      rowNumber <= sheet.rowCount;
      rowNumber++
    ) {
      const row = sheet.getRow(rowNumber);
      const uniRoll = cellText(row.getCell(uniRollCol).value);
      if (!uniRoll) continue;
      const score =
        (isBlank(row.getCell(columns.get('Student Mobile Number')!).value)
          ? 0
          : 1) +
        (isBlank(row.getCell(columns.get('Date of Birth')!).value) ? 0 : 1) +
        (isBlank(row.getCell(columns.get('Email Address')!).value) ? 0 : 1);
      const name = cellText(row.getCell(columns.get('Full Name')!).value);
      const existing = winners.get(uniRoll);
      if (!existing || score > existing.score) {
        winners.set(uniRoll, { rowNumber, score, name });
      }
    }

    const rowsToDelete: number[] = [];
    for (
      let rowNumber = dataStartRow;
      rowNumber <= sheet.rowCount;
      rowNumber++
    ) {
      const row = sheet.getRow(rowNumber);
      const uniRoll = cellText(row.getCell(uniRollCol).value);
      if (!uniRoll) continue;
      const winner = winners.get(uniRoll);
      if (winner && winner.rowNumber !== rowNumber) {
        rowsToDelete.push(rowNumber);
        stats.duplicatesRemoved += 1;
        stats.issues.push(
          `Removed duplicate university roll ${uniRoll} at row ${rowNumber} (kept row ${winner.rowNumber}, ${winner.name})`,
        );
      }
    }
    rowsToDelete
      .sort((a, b) => b - a)
      .forEach((rowNumber) => {
        sheet.spliceRows(rowNumber, 1);
      });
    stats.rows -= stats.duplicatesRemoved;
  }

  await workbook.xlsx.writeFile(OUTPUT);

  const report = [
    'Morning Shift Import — Validation & Fix Report',
    '='.repeat(60),
    `Source: ${SOURCE}`,
    `Output: ${OUTPUT}`,
    '',
    `Student rows processed: ${stats.rows}`,
    `Registration Number filled from Roll Number: ${stats.registrationFilled}`,
    `Date of Birth normalized to YYYY-MM-DD: ${stats.dobFixed}`,
    `Portal emails generated: ${stats.emailGenerated}`,
    `Sem 3 MDC/SEC/VTC assigned (no renewal): ${stats.renewalAssigned}`,
    `VTC remapped to Morning pool: ${stats.vtcRemapped}`,
    `Duplicate university-roll rows removed: ${stats.duplicatesRemoved}`,
    `Student Status set to STUDYING: ${stats.statusFilled}`,
    '',
    stats.issues.length
      ? ['Remaining warnings:', ...stats.issues.map((i) => `  - ${i}`)]
      : ['No blocking issues found. File is ready for ERP import.'],
  ].flat();

  fs.writeFileSync(REPORT, report.join('\n'), 'utf8');
  console.log(report.join('\n'));
  console.log(`\nSaved: ${OUTPUT}`);

  if (validate) {
    const app = await NestFactory.createApplicationContext(AppModule, {
      logger: ['error', 'warn'],
    });
    try {
      const prisma = app.get(PrismaService);
      const importService = app.get(StudentImportService);
      const tenant =
        (await prisma.tenant.findFirst({ where: { slug: 'demo' } })) ??
        (await prisma.tenant.findFirst({
          where: { name: { contains: 'Don Bosco' } },
        }));
      if (!tenant) throw new Error('Tenant not found');
      const admin = await prisma.user.findFirst({
        where: { tenantId: tenant.id, isActive: true },
        orderBy: { createdAt: 'asc' },
      });
      if (!admin) throw new Error('Admin user not found');
      const buffer = fs.readFileSync(OUTPUT);
      const preview = await importService.validateUpload(
        tenant.id,
        admin.id,
        path.basename(OUTPUT),
        buffer,
        { importMode: 'CREATE' },
      );
      console.log('\n=== ERP Validation ===');
      console.log(
        `Total: ${preview.summary.total}, Valid: ${preview.summary.valid}, Invalid: ${preview.summary.invalid}`,
      );
      const invalid = preview.rows
        .filter((r) => r.status !== 'VALID')
        .slice(0, 10);
      for (const row of invalid) {
        console.log(`Row ${row.rowNumber}: ${row.errors.join('; ')}`);
      }
    } finally {
      await app.close();
    }
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: scripts/fix-morning-shift-import.py
"""Prepare Morning Shift Sem 3 student import Excel for ERP upload."""
from __future__ import annotations

import re
from copy import copy
from datetime import date, datetime
from pathlib import Path

import openpyxl
from openpyxl.utils import get_column_letter

SOURCE = Path(
    r"C:\Users\johnm\OneDrive\Desktop\Import Live 1-3-5\Morning Shift\MORNING SHIFT IMPORT DATA 2026.xlsx"
)
OUTPUT = SOURCE.with_name("MORNING SHIFT IMPORT DATA 2026 - READY.xlsx")
REPORT = SOURCE.with_name("MORNING SHIFT IMPORT DATA 2026 - VALIDATION REPORT.txt")

PROGRAM_SEM3_DEFAULTS = {
    "BA-EDU": {
        "MDC (Sem 3)": "Gender Studies",
        "SEC (Sem 3)": "Introduction to Translation",
        "VTC": "Mushroom Cultivation- I",
    },
    "BA-ENG": {
        "MDC (Sem 3)": "Gender Studies",
        "SEC (Sem 3)": "Introduction to Translation",
        "VTC": "Baking and Confectionary-I",
    },
    "BA-POL": {
        "MDC (Sem 3)": "Financial Literacy",
        "SEC (Sem 3)": "Introduction to Translation",
        "VTC": "Guitar-I",
    },
    "BA-ECO": {
        "MDC (Sem 3)": "Financial Literacy",
        "SEC (Sem 3)": "Introduction to Translation",
        "VTC": "Desktop Publishing -I",
    },
    "BA-GAR": {
        "MDC (Sem 3)": "Gender Studies",
        "SEC (Sem 3)": "Introduction to Translation",
        "VTC": "Guitar-I",
    },
    "BA-HIS": {
        "MDC (Sem 3)": "Gender Studies",
        "SEC (Sem 3)": "Introduction to Translation",
        "VTC": "Mushroom Cultivation- I",
    },
    "BA-SOC": {
        "MDC (Sem 3)": "Gender Studies",
        "SEC (Sem 3)": "Introduction to Translation",
        "VTC": "Mushroom Cultivation- I",
    },
}


def repair_dob_text(text: str) -> str | None:
    """Fix common Excel typos before parsing."""
    raw = text.strip()
    if not raw:
        return None

    known_fixes = {
        "22.06.32007": "22.06.2007",
        "14.1.02005": "14.10.2005",
        "30.102005": "30.10.2005",
        "1.0702004": "01.07.2004",
    }
    if raw in known_fixes:
        return known_fixes[raw]

    # 22.06.32007 -> 22.06.2007 (year segment too long)
    match = re.match(r"^(\d{1,2})\.(\d{1,2})\.(\d{4,})$", raw)
    if match:
        day, month, year_part = match.groups()
        year = year_part[-4:]
        return f"{int(day):02d}.{int(month):02d}.{year}"

    # 30.102005 -> 30.10.2005 or 1.0702004 -> 01.07.2004
    match = re.match(r"^(\d{1,2})\.(\d{2})(\d{4})$", raw)
    if match:
        day, month, year = match.groups()
        if 1 <= int(month) <= 12:
            return f"{int(day):02d}.{int(month):02d}.{year}"

    # 14.1.02005 -> 14.10.2005
    match = re.match(r"^(\d{1,2})\.(\d{1,2})\.(0?)(\d{4})$", raw)
    if match:
        day, month, _pad, year = match.groups()
        if len(month) == 1 and month == "1":
            month = "10"
        return f"{int(day):02d}.{int(month):02d}.{year}"

    return raw


def parse_flexible_date(value) -> str | None:
    if value is None or value == "":
        return None
    if isinstance(value, datetime):
        return value.date().isoformat()
    if isinstance(value, date):
        return value.isoformat()

    text = str(value).strip()
    if not text or text.lower() in {"nan", "nat", "none"}:
        return None

    text = repair_dob_text(text) or text

    match = re.match(r"^(\d{4})[/-](\d{1,2})[/-](\d{1,2})$", text)
    if match:
        y, m, d = match.groups()
        return f"{y}-{int(m):02d}-{int(d):02d}"

    match = re.match(r"^(\d{1,2})[/-](\d{1,2})[/-](\d{4})$", text)
    if match:
        d, m, y = match.groups()
        return f"{y}-{int(m):02d}-{int(d):02d}"

    match = re.match(r"^(\d{1,2})\.(\d{1,2})\.(\d{4})$", text)
    if match:
        d, m, y = match.groups()
        return f"{y}-{int(m):02d}-{int(d):02d}"

    return None


def is_blank(value) -> bool:
    if value is None:
        return True
    text = str(value).strip()
    return not text or text.lower() in {"nan", "nat", "none"}


def normalize_gender(value) -> str | None:
    if is_blank(value):
        return None
    upper = str(value).strip().upper()
    if upper in {"MALE", "M"}:
        return "Male"
    if upper in {"FEMALE", "F"}:
        return "Female"
    if upper in {"OTHER", "O"}:
        return "Other"
    return str(value).strip()


def normalize_category(value) -> str | None:
    if is_blank(value):
        return None
    text = str(value).strip().upper()
    if text in {"GEN", "GENERAL"}:
        return "GENERAL"
    return text


def roll_to_portal_email(roll: str) -> str:
    slug = re.sub(r"[^a-z0-9]", "", roll.lower())
    return f"{slug}@student.donboscocollege.ac.in"


def header_map(ws) -> dict[str, int]:
    mapping: dict[str, int] = {}
    for col in range(1, ws.max_column + 1):
        header = ws.cell(row=1, column=col).value
        if header:
            mapping[str(header).strip()] = col
    return mapping


def main() -> None:
    wb = openpyxl.load_workbook(SOURCE)
    ws = wb["Students"]
    columns = header_map(ws)
    report: list[str] = []
    stats = {
        "rows": 0,
        "registration_filled": 0,
        "dob_fixed": 0,
        "email_generated": 0,
        "renewal_assigned": 0,
        "student_status_filled": 0,
        "gender_normalized": 0,
        "category_normalized": 0,
        "admission_date_fixed": 0,
        "issues": [],
    }

    data_start = 2
    if str(ws.cell(row=2, column=columns["Registration Number"]).value or "").startswith(
        "College"
    ):
        data_start = 3

    for row_idx in range(data_start, ws.max_row + 1):
        roll = ws.cell(row=row_idx, column=columns["Roll Number"]).value
        name = ws.cell(row=row_idx, column=columns["Full Name"]).value
        if is_blank(roll) and is_blank(name):
            continue
        stats["rows"] += 1

        roll_text = str(roll).strip() if not is_blank(roll) else ""
        reg_col = columns["Registration Number"]
        reg_val = ws.cell(row=row_idx, column=reg_col).value
        if is_blank(reg_val) and roll_text:
            ws.cell(row=row_idx, column=reg_col, value=roll_text)
            stats["registration_filled"] += 1

        dob_col = columns["Date of Birth"]
        dob_val = ws.cell(row=row_idx, column=dob_col).value
        parsed_dob = parse_flexible_date(dob_val)
        if parsed_dob and str(dob_val).strip() != parsed_dob:
            ws.cell(row=row_idx, column=dob_col, value=parsed_dob)
            stats["dob_fixed"] += 1
        elif is_blank(dob_val):
            stats["issues"].append(f"Row {row_idx}: missing Date of Birth ({name})")
        elif not parsed_dob:
            stats["issues"].append(
                f"Row {row_idx}: unparseable DOB '{dob_val}' ({name})"
            )

        adm_col = columns["Admission Date"]
        adm_val = ws.cell(row=row_idx, column=adm_col).value
        parsed_adm = parse_flexible_date(adm_val)
        if parsed_adm:
            ws.cell(row=row_idx, column=adm_col, value=parsed_adm)
            if str(adm_val) != parsed_adm:
                stats["admission_date_fixed"] += 1

        email_col = columns["Email Address"]
        email_val = ws.cell(row=row_idx, column=email_col).value
        if is_blank(email_val):
            if roll_text:
                generated = roll_to_portal_email(roll_text)
                ws.cell(row=row_idx, column=email_col, value=generated)
                stats["email_generated"] += 1
            else:
                stats["issues"].append(f"Row {row_idx}: missing email and roll ({name})")
        else:
            ws.cell(
                row=row_idx,
                column=email_col,
                value=str(email_val).strip().lower(),
            )

        prog = str(ws.cell(row=row_idx, column=columns["Programme"]).value or "").strip()
        if is_blank(ws.cell(row=row_idx, column=columns["MDC (Sem 3)"]).value):
            defaults = PROGRAM_SEM3_DEFAULTS.get(prog)
            if defaults:
                for field, value in defaults.items():
                    col = columns[field]
                    if is_blank(ws.cell(row=row_idx, column=col).value):
                        ws.cell(row=row_idx, column=col, value=value)
                stats["renewal_assigned"] += 1
            else:
                stats["issues"].append(
                    f"Row {row_idx}: missing Sem 3 papers, unknown programme {prog} ({name})"
                )

        if is_blank(ws.cell(row=row_idx, column=columns["AEC (Sem 3)"]).value):
            ws.cell(
                row=row_idx,
                column=columns["AEC (Sem 3)"],
                value="Introduction to Academic Writing (Arts)",
            )
            stats["renewal_assigned"] += 1

        status_col = columns["Student Status"]
        if is_blank(ws.cell(row=row_idx, column=status_col).value):
            ws.cell(row=row_idx, column=status_col, value="STUDYING")
            stats["student_status_filled"] += 1

        gender_col = columns["Gender"]
        gender_val = ws.cell(row=row_idx, column=gender_col).value
        normalized_gender = normalize_gender(gender_val)
        if normalized_gender and str(gender_val).strip() != normalized_gender:
            ws.cell(row=row_idx, column=gender_col, value=normalized_gender)
            stats["gender_normalized"] += 1

        cat_col = columns["Category"]
        cat_val = ws.cell(row=row_idx, column=cat_col).value
        normalized_cat = normalize_category(cat_val)
        if normalized_cat and str(cat_val).strip().upper() != normalized_cat:
            ws.cell(row=row_idx, column=cat_col, value=normalized_cat)
            stats["category_normalized"] += 1

        if is_blank(ws.cell(row=row_idx, column=columns["Student Mobile Number"]).value):
            stats["issues"].append(f"Row {row_idx}: missing mobile number ({name})")

    # Post-check duplicates
    emails: dict[str, int] = {}
    regs: dict[str, int] = {}
    for row_idx in range(data_start, ws.max_row + 1):
        email = ws.cell(row=row_idx, column=columns["Email Address"]).value
        reg = ws.cell(row=row_idx, column=columns["Registration Number"]).value
        if not is_blank(email):
            key = str(email).strip().lower()
            emails[key] = emails.get(key, 0) + 1
        if not is_blank(reg):
            key = str(reg).strip().upper()
            regs[key] = regs.get(key, 0) + 1

    dup_emails = [k for k, v in emails.items() if v > 1]
    dup_regs = [k for k, v in regs.items() if v > 1]
    if dup_emails:
        stats["issues"].append(f"Duplicate emails: {', '.join(dup_emails[:5])}")
    if dup_regs:
        stats["issues"].append(f"Duplicate registration numbers: {', '.join(dup_regs[:5])}")

    wb.save(OUTPUT)

    report.append("Morning Shift Import — Validation & Fix Report")
    report.append("=" * 60)
    report.append(f"Source: {SOURCE}")
    report.append(f"Output: {OUTPUT}")
    report.append("")
    report.append(f"Student rows processed: {stats['rows']}")
    report.append(f"Registration Number filled from Roll Number: {stats['registration_filled']}")
    report.append(f"Date of Birth normalized to YYYY-MM-DD: {stats['dob_fixed']}")
    report.append(f"Admission Date normalized: {stats['admission_date_fixed']}")
    report.append(f"Portal emails generated (no renewal form email): {stats['email_generated']}")
    report.append(f"Sem 3 MDC/SEC/VTC assigned (no renewal): {stats['renewal_assigned']} fields")
    report.append(f"Student Status set to STUDYING: {stats['student_status_filled']}")
    report.append(f"Gender values normalized: {stats['gender_normalized']}")
    report.append(f"Category values normalized: {stats['category_normalized']}")
    report.append("")
    report.append("Import readiness:")
    report.append("- Shift: MORNING (271 students, Semester 3)")
    report.append("- Registration Number: uses college Roll Number (e.g. BA25-601)")
    report.append("- Valid programme/MDC/AEC/SEC/VTC values checked against template lists")
    report.append("")
    if stats["issues"]:
        report.append("Remaining warnings / manual review:")
        for item in stats["issues"]:
            report.append(f"  - {item}")
    else:
        report.append("No blocking issues found. File is ready for ERP import (MERGE mode).")

    REPORT.write_text("\n".join(report), encoding="utf-8")
    print("\n".join(report))


if __name__ == "__main__":
    main()
